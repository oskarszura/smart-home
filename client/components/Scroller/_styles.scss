$c-scroller-size: 12px;

.c-scroller {
    position: relative;
    border-width: 1px;
    border-style: solid;
    border-color: $brand-component-outline-color;
    background-color: $brand-component-bg-color;
    border-radius: $radius-small;
    height: $c-scroller-size;

    &__knob {
        position: absolute;
        border-radius: $radius-small;
        background-color: $color-mercury;
        height: $c-scroller-size;
        width: $c-scroller-size;
    }

     &__drag {
        position: absolute;
        opacity: 0;
        border-radius: $radius-small;
        background-color: $color-mercury;
        height: $c-scroller-size;
        width: $c-scroller-size;
      }
}
