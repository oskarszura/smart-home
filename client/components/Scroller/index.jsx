// @flow
import _ from 'lodash';
import React from 'react';
import ReactDOM from 'react-dom';

const KNOB_SIZE = 12;
const BORDER_SIZE = 1;

type Props = {
  className?: string,
  onScrollChange: (value: number) => void,
  min: number,
  max: number,
};

type State = {
  value: number,
};

class Scroller extends React.PureComponent<Props, State> {
  static defaultProps = {
    className: '',
  };

  constructor() {
    super();
    this.state = {
      value: 0,
    };
  }

  scroll(event: SyntheticDragEvent<Element>) {
    const { clientX } = event;
    const { min, max } = this.props;

    /* eslint-disable react/no-find-dom-node */
    const node: null | Element | Text = ReactDOM.findDOMNode(this);
    /* eslint-enable react/no-find-dom-node */
    if (node === null || node instanceof Text) {
      return;
    }
    const rect: ClientRect = node.getBoundingClientRect();
    const mappedX = clientX - rect.left;
    const maxWidth = rect.width;
    const scaledX = (mappedX / maxWidth) * (max - min);

    this.move(mappedX, scaledX, maxWidth);
  }

  onMove(event: SyntheticDragEvent<Element>) {
    event.preventDefault();
    this.scroll(event);
  }

  onDragEnd(event: SyntheticDragEvent<Element>) {
    event.preventDefault();
    this.scroll(event);
  }

  move(x: number, vx: number, maxWidth: number) {
    const { onScrollChange } = this.props;
    const value = _.min([
      _.max([x, 0]),
      maxWidth - KNOB_SIZE - 2 * BORDER_SIZE,
    ]);

    onScrollChange(vx);
    this.setState({
      value,
    });
  }

  render() {
    const { className } = this.props;
    const { value } = this.state;
    const classes = `c-scroller ${className}`;

    return (
      <div className={classes}>
        <div
          className="c-scroller__knob"
          style={{
            left: value,
          }}
        />
        <div
          className="c-scroller__drag"
          draggable
          onDrag={(event: SyntheticDragEvent<HTMLDivElement>) => {
            this.onMove(event);
          }}
          onDragEnd={(event: SyntheticDragEvent<HTMLDivElement>) => {
            this.onDragEnd(event);
          }}
          style={{
            left: value,
          }}
        />
      </div>
    );
  }
}

export default Scroller;
