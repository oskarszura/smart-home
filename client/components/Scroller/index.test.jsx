import _ from 'lodash';
import React from 'react';
import { mount } from 'enzyme';
import Scroller from './index';

describe('Scroller', () => {
  it('should render correctly', () => {
    const component = mount(
      <Scroller onScrollChange={_.noop} min={10} max={100} />
    );

    expect(component).toMatchSnapshot();
  });
});
