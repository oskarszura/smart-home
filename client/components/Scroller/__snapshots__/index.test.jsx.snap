// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scroller should render correctly 1`] = `
<Scroller
  className=""
  max={100}
  min={10}
  onScrollChange={[Function]}
>
  <div
    className="c-scroller "
  >
    <div
      className="c-scroller__knob"
      style={
        Object {
          "left": 0,
        }
      }
    />
    <div
      className="c-scroller__drag"
      draggable={true}
      onDrag={[Function]}
      onDragEnd={[Function]}
      style={
        Object {
          "left": 0,
        }
      }
    />
  </div>
</Scroller>
`;
